--- matrixssl-3-7-2b-open-original/core/coreConfig.h	2015-03-18 20:57:05.000000000 +0000
+++ matrixssl-3-7-2b-open/core/coreConfig.h	2016-02-24 21:04:19.000000000 +0000
@@ -50,7 +50,6 @@
 */
 /* #define USE_CORE_TRACE */
 
-
 /******************************************************************************/
 /*
 	Include the osdepMutex family of APIs
@@ -59,6 +58,17 @@
 
 /******************************************************************************/
 /*
+  Share the session cache between processes - note that this is using annonymous
+  shared memory - so the memory will be available only between fork'ed processes
+*/
+/* #define USE_SHARED_SESSION_CACHE */
+
+#if defined(USE_SHARED_SESSION_CACHE) && !defined(USE_MULTITHREADING)
+#error USE_MULTITHREADING must be enabled when using shared session cache (synchronization mechanism)
+#endif
+
+/******************************************************************************/
+/*
 	Use C_GenerateRandom for entropy gathering
 */
 /* #define USE_PKCS11_ENTROPY */
