--- matrixssl-3-7-2b-open-original/core/WIN32/osdep.c	2015-03-18 20:57:05.000000000 +0000
+++ matrixssl-3-7-2b-open/core/WIN32/osdep.c	2016-03-30 16:04:02.000000000 +0000
@@ -68,13 +68,13 @@ int32 psGetTime(psTime_t *t, void *userP
 	if (t == NULL) {
 		QueryPerformanceCounter(&lt);
 		diff = lt.QuadPart - hiresStart.QuadPart;
-		d = (int32)((diff * 1000) / hiresFreq.QuadPart);
+		d = (int32)(diff / hiresFreq.QuadPart);
 		return d;
 	}
 
 	QueryPerformanceCounter(t);
 	diff = t->QuadPart - hiresStart.QuadPart;
-	d = (int32)((diff * 1000) / hiresFreq.QuadPart);
+	d = (int32)(diff / hiresFreq.QuadPart);
 	return d;
 }
 
