--- matrixssl-3-7-2b-open-original/core/corelib.c	2015-03-18 20:57:05.000000000 +0000
+++ matrixssl-3-7-2b-open/core/corelib.c	2016-02-29 15:42:03.000000000 +0000
@@ -311,4 +311,11 @@ void psBurnStack(uint32 len)
 }
 
 /******************************************************************************/
-
+/*
+	All memory allocated by this lib must also be freed to avoid applications
+	that redefine malloc/free (see Perl)
+*/
+void psFreeBuffer(void *ptr)
+{
+	free(ptr);
+}
