--- matrixssl-3-7-2b-open-original/core/POSIX/osdep.c	2015-03-18 20:57:05.000000000 +0000
+++ matrixssl-3-7-2b-open/core/POSIX/osdep.c	2016-02-24 21:04:19.000000000 +0000
@@ -250,6 +250,15 @@ int osdepMutexOpen(void)
 		psErrorInt("pthread_mutexattr_init failed %d\n", rc);
 		return PS_PLATFORM_FAIL;
 	}
+
+#ifdef USE_SHARED_SESSION_CACHE
+	// make the mutex process shareable
+	if ((rc = pthread_mutexattr_setpshared(&attr, PTHREAD_PROCESS_SHARED)) < 0) {
+		psErrorInt("pthread_mutexattr_setpshared failed %d\n", rc);
+		return PS_PLATFORM_FAIL;
+	}
+#endif
+
 	return PS_SUCCESS;
 }
 
