--- matrixssl-3-7-2b-open-original/crypto/keyformat/x509.c	2015-03-18 20:57:05.000000000 +0000
+++ matrixssl-3-7-2b-open/crypto/keyformat/x509.c	2016-02-24 21:04:19.000000000 +0000
@@ -2302,8 +2302,13 @@ static int32 validateDateRange(psX509Cer
 #else
 #warning "CERTIFICATE DATE VALIDITY NOT SUPPORTED ON THIS PLATFORM."
 #endif
-	cert->authFailFlags |= PS_CERT_AUTH_FAIL_DATE_FLAG;
+
+#ifdef ALL_CERTS_PASS_DATE_VALIDATION
 	return 0;
+#else
+	cert->authFailFlags |= PS_CERT_AUTH_FAIL_DATE_FLAG;
+#endif
+
 #endif /* POSIX */
 }
 
@@ -3026,7 +3031,12 @@ int32 psX509AuthenticateCert(psPool_t *p
 
 		/* Ensure keyCertSign of KeyUsage. The second byte of the BIT STRING
 			will always contain the relevant information. */
-		if ( ! (ic->extensions.keyUsageFlags & KEY_USAGE_KEY_CERT_SIGN)) {
+		
+		// Added constrain to perform this check only if the issuer certificate is not the last
+		// (self-signed) certificate in the chain. By design, matrixSSL duplicates the self-signed
+		// last CA entry. The caller of this function expects the last entry in the chain to fail
+		// this test but not the penultimate one
+		if ((ic->next != NULL) && (ic != ic->next) && !(ic->extensions.keyUsageFlags & KEY_USAGE_KEY_CERT_SIGN)) {
 			/* @security If keyUsageFlags is zero, it may not exist at all
 				in the cert. This is allowed if the cert was issued before
 				the RFC was updated to require this field for CA certificates.
