--- matrixssl-3-7-2b-open-original/crypto/math/pstm_montgomery_reduce.c	2015-03-18 20:57:05.000000000 +0000
+++ matrixssl-3-7-2b-open/crypto/math/pstm_montgomery_reduce.c	2016-02-24 21:04:19.000000000 +0000
@@ -61,7 +61,7 @@ asm(                                    
    "movl %%edx,%1 \n\t"                                   \
 :"=g"(_c[LO]), "=r"(cy)                                   \
 :"0"(_c[LO]), "1"(cy), "g"(mu), "g"(*tmpm++)              \
-: "%eax", "%edx", "%cc")
+: "%eax", "%edx", "cc")
 
 #define PROPCARRY                           \
 asm(                                        \
@@ -70,7 +70,7 @@ asm(                                    
    "movzbl %%al,%1 \n\t"                    \
 :"=g"(_c[LO]), "=r"(cy)                     \
 :"0"(_c[LO]), "1"(cy)                       \
-: "%eax", "%cc")
+: "%eax", "cc")
 
 /******************************************************************************/
 #elif defined(PSTM_X86_64)
@@ -223,7 +223,7 @@ asm(                                \
 	" STR    r0,%1            \n\t" \
 	:"=r"(cy),"=m"(_c[0])\
 	:"0"(cy),"r"(mu),"r"(*tmpm++),"m"(_c[0])\
-	:"r0","%cc");
+	:"r0","cc");
 #define PROPCARRY                  \
 asm(                               \
 	" LDR   r0,%1            \n\t" \
@@ -234,7 +234,7 @@ asm(                               \
 	" MOVCC %0,#0            \n\t" \
 	:"=r"(cy),"=m"(_c[0])\
 	:"0"(cy),"m"(_c[0])\
-	:"r0","%cc");
+	:"r0","cc");
 #else /* Non-Thumb2 code */
 //#pragma message ("Using 32 bit ARM Assembly Optimizations")
 #define INNERMUL                    \
@@ -247,7 +247,7 @@ asm(                                \
 	" STR    r0,%1            \n\t" \
 	:"=r"(cy),"=m"(_c[0])\
 	:"0"(cy),"r"(mu),"r"(*tmpm++),"m"(_c[0])\
-	:"r0","%cc");
+	:"r0","cc");
 #define PROPCARRY                  \
 asm(                               \
 	" LDR   r0,%1            \n\t" \
@@ -257,7 +257,7 @@ asm(                               \
 	" MOVCC %0,#0            \n\t" \
 	:"=r"(cy),"=m"(_c[0])\
 	:"0"(cy),"m"(_c[0])\
-	:"r0","%cc");
+	:"r0","cc");
 #endif /* __thumb2__ */
 
 
