--- matrixssl-3-7-2b-open-original/crypto/pubkey/pkcs.c	2015-03-18 20:57:05.000000000 +0000
+++ matrixssl-3-7-2b-open/crypto/pubkey/pkcs.c	2016-02-24 21:04:19.000000000 +0000
@@ -913,7 +913,9 @@ static int32 pkcs12import(psPool_t *pool
 		memset_s(decryptKey, keyLen, 0x0, keyLen);
 		psFree(decryptKey, pool);
 	}
-	psFree(decryptKey, pool);
+
+	// if decryptKey != NULL then it's freed above
+	// psFree(decryptKey, pool);
 	psFree(iv, pool);
 
 	*plaintext = pt;
